source/extensions/filters/http/jwt_authn/filter_config.h

```diff
@@ -68,8 +68,18 @@ class FilterConfig : public Logger::Loggable<Logger::Id::jwt>, public AuthFactor
         tls_(context.threadLocal().allocateSlot()), cm_(context.clusterManager()),
         time_source_(context.dispatcher().timeSource()), api_(context.api()) {
     ENVOY_LOG(info, "Loaded JwtAuthConfig: {}", proto_config_.DebugString());
-    tls_->set([this](Event::Dispatcher&) -> ThreadLocal::ThreadLocalObjectSharedPtr {
-      return std::make_shared<ThreadLocalCache>(proto_config_, time_source_, api_);
+
+    // note: `this` and `context` has a a lifetime of the listener.
+    // that may be shorter of the tls callback if the listener is torn shortly after it is created.
+    // context.dispatcher().timeSource() and context.api() are from the server context who's
+    // lifetime is longer. so we use them explicitly.
+    // we copy over the proto as we can't guarantee its lifetime otherwise
+    auto& timeSource = context.dispatcher().timeSource();
+    auto& api = context.api();
+    auto proto_config_copy = proto_config_;
+    tls_->set([proto_config_copy, &timeSource,
+               &api](Event::Dispatcher&) -> ThreadLocal::ThreadLocalObjectSharedPtr {
+      return std::make_shared<ThreadLocalCache>(proto_config_copy, timeSource, api);
```

Comment:
proto_config_copy is owned by the callback now,  it will be gone when it is called.  I am not sure if ThreadLocalCache make a copy of it. 

 If not,  please do so in this code:  https://cs.corp.google.com/piper///depot/google3/third_party/envoy/src/source/extensions/filters/http/jwt_authn/jwks_cache.cc?type=cs&rcl=292261389&g=0&l=128

Otherwise,  it may just refer to the config that have been removed.  

The old code is referring to the config owned by the FilterConfig which is safe.

Reply:
from looking at the code, ThreadLocalCache doesn't keep a reference to it after construction, so this should be fine

Reply:
the old cold was referring to the FilterConfig indeed, but that wasn't safe..
The issue is that the lifetime of the FilterConfig is the same as the listener in the main thread, while the callback has an independent lifetime in each of the worker threads.

Reply:
JwtCacheImpl is using reference.

https://cs.corp.google.com/piper///depot/google3/third_party/envoy/src/source/extensions/filters/http/jwt_authn/jwks_cache.cc?rcl=292261389&g=0&l=87

It will be referring to your proto_config_copy which have been deleted after the call.

We should keep a copy in the JwtCacheImpl

Reply:
This may explain why tsan or asan tests failed.  It is access the memory after free.

Reply:
right - good catch, i'll fix.